modules/library/coverage/src/main/java/org/geotools/image/ImageWorker.java

```
@@ -3859,18 +3866,52 @@ public ImageWorker affine(AffineTransform tx, Interpolation interpolation, doubl
                             paramBlk.add(warp).add(interpolation).add(bgValues);
                             paramBlk.set(nodata, 4);
-                            if(isNoDataNeeded() && bgValues != null && bgValues.length > 0){
+                            // Try to reproject ROI after Warp
+                            ROI newROI = null;
+                            if(roi != null){
+                                ROI reprojectedROI = roi;
+                                try{
+                                    MathTransform inverse = originalTransform.inverse();
+                                    if(inverse instanceof AffineTransform){
+                                        AffineTransform inv = (AffineTransform)inverse;
+                                        newROI = reprojectedROI.transform(inv);
+                                    }
+                                }catch(Exception e){
+                                    if(LOGGER.isLoggable(Level.WARNING)){
+                                        LOGGER.log(Level.WARNING, "Unable to compute the inverse of the new ROI provided", e);
+                                    }
+                                }
+                            }
+  
+                            if(newROI != null){
+                                setROI(newROI);
+                                paramBlk.set(newROI, 3);
+                            }
+                            if((isNoDataNeeded() || newROI != null) && bgValues != null && bgValues.length > 0){
                                 setNoData(RangeFactory.create(bgValues[0], bgValues[0]));
                             }
                         } else {
                             // Intersect ROIs
                             ROI newROI = null;
                             if (roi != null) {
-                                newROI = roi.intersect((ROI) property);
+                                // Try to reproject ROI after Warp
```

Same as above, formatting off, risk of silently losing ROI